---
title: "Homework 6"
author: 'Roberto Villegas-Diaz'
collaborators: 'Roberto Villegas-Diaz'
header-includes:
- \usepackage{caption}
- \usepackage{fancyhdr}
- \usepackage{float}
- \usepackage{lastpage}
- \pagestyle{fancy}
- \fancyhead[L]{STAT736 Bioinformatics}
- \newcommand{\ans}{{\color{blue}Answer:}}
output:
  pdf_document: default
#bibliography: bibliography.bib
#nocite: '@*'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(message=F,warning=F,echo=F,fig.width=10, fig.height=5,cache = F)
```

```{r}
# Load libraries
library(limma) # Bioconductor
library(DESeq2) # Bioconductor
library(dplyr)  # R package
library(latex2exp)
```

# 1. Reading data
First we read in counts data inside the `readCounts.txt`.
```{R  message = FALSE, cache = F}
countData = read.csv("readCounts.txt", skip = 1, sep = '\t', header = TRUE) # nrows = 10, 
countData[,2:6] <- NULL
# colnames(countData) <- c("GeneID","SRR2121770","SRR2121771","SRR2121774","SRR2121775",
#                          "SRR2121778","SRR2121779","SRR2121780","SRR2121781",
#                          "SRR2121786","SRR2121787","SRR2121788","SRR2121789")
# colnames(countData) <- c("GeneID","p53.null.Bcells.mock.1","p53.null.Bcells.mock.2","p53.null.Bcells.IR.1",
#                          "p53.null.Bcells.IR.2","p53.wt.Bcells.mock.1","p53.wt.Bcells.mock.2",
#                          "p53.wt.Bcells.mock.3"," p53.wt.Bcells.mock.4","p53.wt.Bcells.IR.1",
#                          "p53.wt.Bcells.IR.2","p53.wt.Bcells.IR.3","p53.wt.Bcells.IR.4")
colnames(countData) <- c("GeneID","null.mock.1","null.mock.2","null.IR.1",
                         "null.IR.2","wt.mock.1","wt.mock.2",
                         "wt.mock.3","wt.mock.4","wt.IR.1",
                         "wt.IR.2","wt.IR.3","wt.IR.4")
GeneID <- countData[,1]
countData[,1] <- NULL
rownames(countData) <- GeneID
```

# 2. Simple exploration of the raw data
```{r cars}
dim(countData)
```


```{r}
summary(countData)
```


```{r, fig.width=10, fig.height=5}
barplot(colSums(countData)/1e6, col="green",las=3,main="Total read counts (millions)", ylab="Total reads in millions")
```

```{r plot,  fig.width=10, fig.height=5}
transformation <- paste0("Log_2(",colnames(countData)[1],")")
hist(log2(countData[,1]),br=200, xlab = TeX(transformation),
     main = TeX(paste0("Histogram of ",transformation)))
```

### Simple log transformation on unnormalized data
```{r,fig.width=10, fig.height=5}
logCountData = log2(1 + countData)
par(mfrow = c(1, 2))  # two columns
hist(logCountData[,1], xlab = TeX(transformation),
     main = TeX(paste0("Histogram of ",transformation)))
boxplot(logCountData,las=3)
```

```{r, fig.width=10, fig.height=5}
  x <- logCountData
	myColors = rainbow(dim(x)[2])
	plot(density(x[,1]),col = myColors[1], lwd=2,
	  xlab="Expresson values", ylab="Density", main= "Distribution of transformed data",
	  ylim=c(0, max(density(x[,1])$y)+.02 ) )
	  
	for( i in 2:dim(x)[2] )
	lines(density(x[,i]),col=myColors[i], lwd=2 )
    legend("topright", cex=1.1,colnames(x), lty=rep(1,dim(x)[2]), col=myColors )	
```

```{r, fig.width=10, fig.height=5}
xlab <- paste0("Log_2(",colnames(countData)[1],")")
ylab <- paste0("Log_2(",colnames(countData)[2],")")
plot(logCountData[,1],logCountData[,2],
     xlab = TeX(xlab),
     ylab = TeX(ylab),
     main = TeX(paste0(ylab, " vs ",xlab)))
```


# 3. Filtering, normalization, and transformation using DESeq2
## Define sample groups
First we define a function for parsing samples into groups.
Define sample groups based on column names
``` {r}
detectGroups <- function (x){  # x are col names
	tem <- gsub(".[0-9]*$","",x) # Remove all numbers from end
	#tem = gsub("_Rep|_rep|_REP","",tem)
	tem <- gsub("_$","",tem); # remove "_" from end
	tem <- gsub("_Rep$","",tem); # remove "_Rep" from end
	tem <- gsub("_rep$","",tem); # remove "_rep" from end
	tem <- gsub("_REP$","",tem)  # remove "_REP" from end
 	return(tem)
}

#groups = as.character(detectGroups(colnames(countData)))
groups <- data.frame(t(strsplit2(as.character(detectGroups(colnames(countData))),'[.]')))
colnames(groups) <- colnames(countData)
rownames(groups) <- c("p53","treatment")
```

<!-- ### Make sample info data frame.  -->
<!-- It can be more complex, think this with multiple columns for using in the model. -->
<!-- ```{r} -->
<!-- colData = cbind(colnames(countData), groups) -->
<!-- colData -->
<!-- ``` -->


Now we need to add sample IDs, as this is very likely a paired situation. We add a new column this is according to this [post](https://support.bioconductor.org/p/84241/) 
```{r}
# colData = as.data.frame(cbind(colnames(countData), groups))
# sampleID = as.factor(c(1:4,1:4))
# sampleID = as.factor(c(1,1,2,2,3,3,3,3,4,4,4,4))
# colData = cbind(data.frame(colData),sampleID)
colData <- data.frame(groups = as.character(detectGroups(colnames(countData))), 
                      t(groups))
colData
str(colData)
```

### Set up the DESeqDataSet Object and run the DESeq pipeline

```{r message = FALSE }
dds = DESeqDataSetFromMatrix(countData = countData,
                             colData = colData,
                             design = ~ p53 + treatment)   # note that the study design is changed.
dds = DESeq(dds)  # main function
nrow(dds)
```

### Filtering 
```{r}
dds <- dds[ rowSums(counts(dds)) > 5, ]
nrow(dds)
```

## rlog transformation
```{r}
rld <- rlog(dds, blind = FALSE)
head(assay(rld), 3)
```

### VSD transformation
```{r}
vsd <- vst(dds, blind = FALSE)
head(assay(vsd), 3)
```

For the log2 approach, we need to first estimate size factors to account for sequencing depth, and then specify normalized=TRUE. Sequencing depth correction is done automatically for the rlog and the vst.

```{r}
dds <- estimateSizeFactors(dds)
sizeFactors(dds)
```

### Started log on scaled data
Usings the normalized=TRUE option in the counts( ) method of DESeq2, we adjust for different library sizes.
```{r}
slog <- log2(counts(dds, normalized=TRUE)+1)
head(slog)
```

```{r}
par(mfrow = c(1, 3))  # 3 columns
plot(slog[,1],slog[,2])
plot(assay(rld)[,1],assay(rld)[,2])
plot(assay(vsd)[,1],assay(vsd)[,2])
```


```{r}
par(mfrow = c(1, 3))  # 3 columns
slog <- log2(counts(dds, normalized=TRUE)+1)
plot(slog[,1],slog[,2])
slog <- log2(counts(dds, normalized=TRUE)+4)
plot(slog[,1],slog[,2], xlim=c(0,20))
slog <- log2(counts(dds, normalized=TRUE)+20)
plot(slog[,1],slog[,2],xlim=c(0,20))
```

A more elegant plot using ggplot2 according to DESeq2 tutorial:

```{r}
library("dplyr")
library("ggplot2")

df <- bind_rows(
  as_tibble(slog[,1:2]) %>%
         mutate(transformation = "log2(x + 1)"),
  as_tibble(assay(rld)[, 1:2]) %>% mutate(transformation = "rlog"),
  as_tibble(assay(vsd)[, 1:2]) %>% mutate(transformation = "vst"))
  
colnames(df)[1:2] <- c("x", "y")  

ggplot(df, aes(x = x, y = y)) + geom_hex(bins = 80) +
  coord_fixed() + facet_grid( . ~ transformation)

```

# 4. Exploratory Data Analysis
## PCA plot    
```{r, message=FALSE}
plotPCA(rld, intgroup = c("groups"))
```

The figure looks odd. What we need is to make x and y aspect ratios independent of the data ranges. 
<!-- After some googling, I landed on this [page.](https://stackoverflow.com/questions/31939838/ggplot2-aspect-ratio-overpowers-coord-equal-or-coord-fixed) -->

```{r}
plotPCA(rld, intgroup = c("groups")) + theme(aspect.ratio=1)
```

You can do the PCA plot yourself by building a plot using ggplot2, step by step

```{r}
pca.object <- prcomp(t(assay(rld))) # PCA 
pcaData = as.data.frame(pca.object$x[,1:2]); 
pcaData = cbind(pcaData,detectGroups(colnames(assay(rld)) ))
colnames(pcaData) = c("PC1", "PC2", "Type")
percentVar=round(100*summary(pca.object)$importance[2,1:2],0)
#plot
p=ggplot(pcaData, aes(PC1, PC2, color=Type, shape = Type)) + geom_point(size=5) 
p=p+xlab(paste0("PC1: ",percentVar[1],"% variance")) 
p=p+ylab(paste0("PC2: ",percentVar[2],"% variance")) 
p=p+ggtitle("Principal component analysis (PCA)")+coord_fixed(ratio=1.0)+ 
    theme(plot.title = element_text(size = 16,hjust = 0.5)) + theme(aspect.ratio=1) +
    theme(axis.text.x = element_text( size = 16),
    axis.text.y = element_text( size = 16),
    axis.title.x = element_text( size = 16),
    axis.title.y = element_text( size = 16) ) +
  theme(legend.text=element_text(size=16))
print(p)
```

## MDS plot
Similarily, we can produce MDS plots. As you can see from the code below, I recycled the ploting code and didn't even change some of the variable names. 

```{r}
dist2 <- function(x, ...)   # distance function = 1-PCC (Pearson's correlation coefficient)
  as.dist(1-cor(t(x), method="pearson"))

fit = cmdscale( dist2(t(assay(rld))) , eig=T, k=2)
mdsData <- as.data.frame(fit$points[,1:2]); 
mdsData <- cbind(mdsData,detectGroups(colnames(assay(rld))) )
colnames(mdsData) = c("x1", "x2", "Type")
	
p<-ggplot(mdsData, aes(x1, x2, color=Type, shape = Type)) + geom_point(size=5) 
p=p+xlab("Dimension 1") 
p=p+ylab("Dimension 2") 
p=p+ggtitle("Multidimensional scaling (MDS)")+ coord_fixed(ratio=1.)+ 
     theme(plot.title = element_text(hjust = 0.5)) + theme(aspect.ratio=1) +
	 	 theme(axis.text.x = element_text( size = 16),
        axis.text.y = element_text( size = 16),
        axis.title.x = element_text( size = 16),
        axis.title.y = element_text( size = 16) ) +
	   theme(legend.text=element_text(size=16))
print(p)
```

## Hierarchical clustering with heatmap
<!-- : my favoriate! -->
<!-- See here for details on heatmap.2 see this [Stackoverflow.com page.](http://stackoverflow.com/questions/15351575/moving-color-key-in-r-heatmap-2-function-of-gplots-package) -->

```{r heatmap, message = FALSE,fig.width=7,fig.height=7}
library(gplots)

hclust2 <- function(x, method="average", ...)  # average linkage in hierarchical clustering
  hclust(x, method=method, ...)

n=50 # number of top genes by standard deviation

x = assay(rld)
if(n>dim(x)[1]) n = dim(x)[1] # max	as data

x = x[order(apply(x,1,sd),decreasing=TRUE),]  # sort genes by standard deviation

x = x[1:n,]   # only keep the n genes

# this will cutoff very large values, which could skew the color 
x=as.matrix(x[1:n,])-apply(x[1:n,],1,mean)
cutoff = median(unlist(x)) + 4*sd (unlist(x)) 
x[x>cutoff] <- cutoff
cutoff = median(unlist(x)) - 4*sd (unlist(x)) 
x[x< cutoff] <- cutoff
	
groups = detectGroups(colnames(x) )
groups.colors = rainbow(length(unique(groups) ) )


	lmat = rbind(c(5,4),c(0,1),c(3,2))
	lwid = c(1.5,4)
	lhei = c(1,.2,4)

heatmap.2(x, distfun = dist2,hclustfun=hclust2,
	 col=greenred(75), density.info="none", trace="none", scale="none", keysize=.5
	,key=T, symkey=F
	#,labRow=labRow
	,ColSideColors=groups.colors[ as.factor(groups)]
	,margins=c(8,8)
	,cexRow=1
	,srtCol=45
	,cexCol=1.  # size of font for sample names
	,lmat = lmat, lwid = lwid, lhei = lhei
	)
```

## K-means clustering of genes

```{r kmeans}
#library(shiny)
#numericInput("k", "How many cars?", 5)
k <- 5
#renderPlot({
n=100 # number of top genes by standard deviation
x = assay(rld)
if(n>dim(x)[1]) n = dim(x)[1] # max	as data
x = x[order(apply(x,1,sd),decreasing=TRUE),]  # sort genes by standard deviation
x = x[1:n,]   # only keep the n genes

	x = 100* x[1:n,] / apply(x[1:n,],1,function(y) sum(abs(y))) # L1 norm

	set.seed(2)
	#k = input$k  # number of clusters
	
	cl = kmeans(x,k,iter.max = 50)

	hc <- hclust2(dist2(cl$centers-apply(cl$centers,1,mean) )  )
	tem = match(cl$cluster,hc$order) #  new order 
	x = x[order(tem),] ; 	bar = sort(tem)
	
	# heatmap with color bar define gene groups
myheatmap2 <- function (x,bar,n=-1 ) {
	# number of genes to show
	ngenes = as.character( table(bar))
	if(length(bar) >n && n != -1) {ix = sort( sample(1:length(bar),n) ); bar = bar[ix]; x = x[ix,]  }

	# this will cutoff very large values, which could skew the color 
	x=as.matrix(x)-apply(x,1,mean)
	cutoff = median(unlist(x)) + 3*sd (unlist(x)) 
	x[x>cutoff] <- cutoff
	cutoff = median(unlist(x)) - 3*sd (unlist(x)) 
	x[x< cutoff] <- cutoff
	#colnames(x)= detectGroups(colnames(x))
	 heatmap.2(x,  Rowv =F,Colv=F, dendrogram ="none",
	 col=greenred(75), density.info="none", trace="none", scale="none", keysize=.3
	,key=F, labRow = F
	,RowSideColors = mycolors[bar]
	,margins = c(8, 24)
	,srtCol=45
	)

	 legend.text = paste("Cluster ", toupper(letters)[unique(bar)], " (N=", ngenes,")", sep="")
	par(lend = 1)           # square line ends for the color legend
	legend("topright",      # location of the legend on the heatmap plot
		legend = legend.text, # category labels
		col = mycolors,  # color key
		lty= 1,             # line style
		lwd = 10            # line width
	)

}

mycolors = sort(rainbow(20))[c(1,20,10,11,2,19,3,12,4,13,5,14,6,15,7,16,8,17,9,18)] # 20 colors for kNN clusters
myheatmap2(x-apply(x,1,mean), bar,1000)

#})

```



# 5. Differential expression analysis using DESeq2

```{r, message=FALSE}
dds <- DESeq(dds)
res <- results(dds)
res
```


## Applying cutoff
DESeq2 uses the Benjamini-Hochberg (BH) adjustment (Benjamini and Hochberg 1995) as implemented in the base R p.adjust function

```{r}
res <- results(dds, alpha = 0.5, lfcThreshold=.01)
summary(res)
```

##Sort genes by fold change
```{r}
res <- res[order(abs( res$log2FoldChange), decreasing=TRUE),]
head(res)
```

## MA plot
```{r}
DESeq2::plotMA(res,  ylim = c(-5, 5))
```
<!-- HERE -->

DESeq2 also provides shrinked log fold changes:

```{r, eval=F}
library(DESeq2)
resShrink <- lfcShrink(dds, contrast = c("groups","null.IR","wt.IR"), res=res)
DESeq2::plotMA(resShrink,  ylim = c(-5, 5))
```

## Basic volcano plot
```{r}
library(dplyr)
res1 = as.data.frame(res)
# add a new column using the mutate function in dplyr
res1 = mutate(res1, sig=ifelse(res1$padj<0.05, "FDR<0.05", "Not Sig"))
res1[which(abs(res1$log2FoldChange)<0.5),'sig'] <- "Not Sig"


p = ggplot(res1, aes(log2FoldChange, -log10(padj))) +
  geom_point(aes(col=sig)) +
  scale_color_manual(values=c("red", "black"))
p
```

## Ploting counts of selected genes
```{r}
topGene <- rownames(res)[1]
plotCounts(dds, gene = topGene, intgroup=c("groups"))
```


# 6. Annotating genes using Bioconductor
First we need to install Bioconductor base and some Bioconductor packages.

```{r, eval =FALSE}
#source("https://bioconductor.org/biocLite.R") # runs R code on the website, defines biocLite() function
#biocLite()
#biocLite("org.Hs.eg.db")
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install()
BiocManager::install("org.Mm.eg.db")
BiocManager::install("mygene")
```


```{r, message=FALSE, cache = F, eval = F}
# library(AnnotationDbi)
# library(org.Mm.eg.db)
# library(mygene)
# columns(org.Mm.eg.db)
# res$symbol <- gsub(" ","",row.names(res))
# res$entrez <- mapIds(org.Mm.eg.db,
#                      keys= row.names(res),
#                      column="ENTREZID",
#                      keytype="SYMBOL",
#                      multiVals="first")
# write.csv(res, file = "results.csv")
# head(res)
# genes.query <- queryMany(res$symbol, scopes='ensembl_peptide_id', fields=c('entrezgene', 'go'), species='mouse')

library(biomaRt)
mouse = useMart("ensembl", dataset = "mmusculus_gene_ensembl")
genes.list <- gsub(" ","",row.names(res))
genes.list <- gsub(".[0-9]*$","",genes.list) # Drop exon count
G_list <- getBM(filters = "ensembl_gene_id", 
                attributes = c("ensembl_gene_id", "external_gene_name", "description"),
                values = genes.list, mart = mouse)
colnames(G_list)[2] <- "entrez"
# rownames(G_list) <- G_list$ensembl_gene_id
res$symbol <- genes.list
# res$entrez <- NA
# res$entrez[res$symbol == G_list$ensembl_gene_id] <- G_list$external_gene_name

res$entrez <- NULL
res$tmp <- rownames(res)
df <- merge(as.data.frame(res),
            G_list,
            by.x="symbol",
            by.y="ensembl_gene_id")
rownames(df) <- df$tmp
df$tmp <- NULL
df$description <- NULL
# df$external_gene_name <- NULL
# res$entrez <- merge(as.data.frame(res),#data.frame(symbol = genes.list),
#             G_list,
#             by.x="symbol",
#             by.y="ensembl_gene_id", 
#             all = TRUE)$external_gene_name
# df$external_gene_name
#res$entrez <- G_list$external_gene_name
# write.csv(res, file = "results.csv")
write.csv(df, file = "results.csv")
```

That file contains all genes passed filter. Let's output only the significant genes. We also sorted the genes by fold change.

```{r}
res <- read.csv("results.csv")
rownames(res) <- res$X
res$X <- NULL
resSig = as.data.frame(subset(res,padj<0.5) )
resSig = resSig[order(resSig$log2FoldChange,decreasing=TRUE),]
head(resSig)

write.csv(resSig,"SigGenes.csv", row.names = FALSE)
```

## Enhanced volcano plot with gene symbols
```{r}
library(dplyr)
res1 = as.data.frame(res)
# add a new column using the mutate function in dplyr
res1 = mutate(res1, sig=ifelse(res1$padj<0.5, "FDR<0.05", "Not Sig"))
res1[which(abs(res1$log2FoldChange)<1),'sig'] <- "Not Sig"
p = ggplot(res1, aes(log2FoldChange, -log10(pvalue))) +
  geom_point(aes(col=sig)) +
  scale_color_manual(values=c("red", "black"))
p+geom_text(data=filter(res1, padj<0.05), aes(label=symbol))
```


# Further improved volcano plot with ggrepel
```{r}
library(dplyr)
# Install ggrepel package if needed
# install.packages("devtools")
# devtools::install_github("slowkow/ggrepel")
library(ggrepel) # "repels" overlapping text

p+geom_text_repel(data=filter(res1,  abs(log2FoldChange)>5 | padj < 1e-100 ), aes(label=symbol))
```



```{r, eval=FALSE}
library("ReportingTools")
htmlRep <- HTMLReport(shortName="report", title="My report",
                      reportDirectory="./report")
publish(resSig, htmlRep)
url <- finish(htmlRep)
browseURL(url)
```


<!-- # 7. GO Enrichment analysis using GOstats -->
<!-- ## 399 upregulated genes -->
<!-- ```{r,message=FALSE} -->
<!-- library(GO.db) -->
<!-- library(GOstats) -->
<!-- selectedGenes = unique(resSig[resSig$log2FoldChange>0,'entrez'])  # upregulated genes -->
<!-- universeGenes =  unique( mapIds(org.Hs.eg.db, -->
<!--                      keys=gsub(" ","",row.names(countData)),  # this is causing problems for mapping -->
<!--                      column="ENTREZID", -->
<!--                      keytype="SYMBOL", -->
<!--                      multiVals="first") -->
<!--                     ) -->


<!--  hgCutoff <- 0.001 -->
<!--  params <- new("GOHyperGParams", -->
<!--      geneIds=selectedGenes, -->
<!--      universeGeneIds=universeGenes, -->
<!--      annotation="org.Hs.eg.db", -->
<!--      ontology="BP", -->
<!--      pvalueCutoff=hgCutoff, -->
<!--      conditional=FALSE, -->
<!--      testDirection="over") -->

<!-- hgOver <- hyperGTest(params) -->
<!-- summary(hgOver)[1:10,] -->


<!-- ``` -->


<!-- ```{r} -->
<!-- summary(hgOver)[1:10,c("GOBPID","Pvalue","Term")] -->
<!-- ``` -->
<!-- ```{r} -->
<!-- params1 <- params -->
<!-- ontology(params1) <- "CC" -->
<!-- hgOver <- hyperGTest(params1) -->
<!-- summary(hgOver)[1:10,c("GOCCID","Pvalue","Term")] -->
<!-- ``` -->

<!-- ```{r} -->
<!-- params1 <- params -->
<!-- ontology(params1) <- "MF" -->
<!-- hgOver <- hyperGTest(params1) -->
<!-- summary(hgOver)[1:10,c("GOMFID","Pvalue","Term")] -->
<!-- ``` -->
<!-- ## GO Enrichment analysis  of 535 downregulated genes -->
<!-- ```{r,message=FALSE} -->

<!-- selectedGenes = unique(resSig[resSig$log2FoldChange<0,'entrez'])  # upregulated genes -->

<!--  params <- new("GOHyperGParams", -->
<!--      geneIds=selectedGenes, -->
<!--      universeGeneIds=universeGenes, -->
<!--      annotation="org.Hs.eg.db", -->
<!--      ontology="BP", -->
<!--      pvalueCutoff=hgCutoff, -->
<!--      conditional=FALSE, -->
<!--      testDirection="over") -->

<!-- hgOver <- hyperGTest(params) -->
<!-- summary(hgOver)[1:10,c("GOBPID","Pvalue","Term")] -->
<!-- ``` -->


<!-- # 8. Pathway analysis using expression data -->
<!-- ### Pathway analysis using GAGE based on Stephen Turner's [blog.](http://www.gettinggeneticsdone.com/2015/12/tutorial-rna-seq-differential.html) -->
<!-- ```{r} -->
<!-- # bioconductor packages -->
<!-- # source("https://bioconductor.org/biocLite.R"); -->
<!-- # biocLite(c("pathview","gage","gageData")) -->
<!-- library(pathview) -->
<!-- library(gage) -->

<!-- ``` -->
<!-- ## Prepare data -->
<!-- ```{r} -->
<!-- foldchanges = res$log2FoldChange -->
<!-- names(foldchanges) = res$entrez -->
<!-- head(foldchanges) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- library(gageData) -->
<!-- data(go.sets.hs) -->
<!-- data(go.subs.hs) -->
<!-- gobpsets = go.sets.hs[go.subs.hs$BP] -->
<!-- gobpres = gage(foldchanges, gsets=gobpsets, same.dir=TRUE) -->
<!-- lapply(gobpres, head) -->
<!-- ``` -->


<!-- ## KEGG pathways -->
<!-- ```{r} -->
<!-- library(gageData) -->
<!-- data(kegg.sets.hs) -->
<!-- data(sigmet.idx.hs) -->
<!-- kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs] -->
<!-- head(kegg.sets.hs, 3) -->
<!-- # Get the results -->
<!-- keggres = gage(foldchanges, gsets=kegg.sets.hs, same.dir=TRUE) -->

<!-- # Look at both up (greater), down (less), and statatistics. -->
<!-- lapply(keggres, head) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- # Get the pathways -->
<!-- keggrespathways = data.frame(id=rownames(keggres$less), keggres$less) %>% -->
<!--   tbl_df() %>% -->
<!--   filter(row_number()<=5) %>% -->
<!--   .$id %>% -->
<!--   as.character() -->
<!-- keggrespathways -->
<!-- ``` -->

<!-- ```{r} -->
<!-- # Get the IDs. -->
<!-- keggresids = substr(keggrespathways, start=1, stop=8) -->
<!-- keggresids -->
<!-- ``` -->
<!-- ```{r} -->

<!-- # Define plotting function for applying later -->
<!-- plot_pathway = function(pid) pathview(gene.data=foldchanges, pathway.id=pid, species="hsa", new.signature=FALSE) -->

<!-- # plot multiple pathways (plots saved to disk and returns a throwaway list object) -->
<!-- tmp = sapply(keggresids, function(pid) pathview(gene.data=foldchanges, pathway.id=pid, species="hsa")) -->

<!-- ``` -->
